CFI: Support function pointers for trait methods

maurer opened the pull request:
Adds support for both CFI and KCFI for function pointers to trait methods by attaching both concrete and abstract types to functions.

KCFI does this through generation of a `ReifyShim` on any function pointer for a method that could go into a vtable, and keeping this separate from `ReifyShim`s that are *intended* for vtable us by setting a `ReifyReason` on them.

CFI does this by setting both the concrete and abstract type on every instance.

This should land after #123024 or a similar PR, as it diverges the implementation of CFI vs KCFI.

r? @compiler-errors

maurer committed:
CFI: Support function pointers for trait methods

Adds support for both CFI and KCFI for attaching concrete and abstract
types to functions. KCFI does this through generation of `ReifyShim` on
any function pointer that could go in a vtable, and checking the
`ReifyReason` when emitting the instance. CFI does this by attaching
both the concrete and abstract type to every instance.

TypeID codegen tests are switched to be anchored on the left rather than
the right in order to allow emission of additional type attachments.

Fixes #115953

maurer committed:
Track reason for creating a `ReifyShim`

KCFI needs to be able to tell which kind of `ReifyShim` it is examining
in order to decide whether to use a concrete type (`FnPtr` case) or an
abstract case (`Vtable` case). You can *almost* tell this from context,
but there is one case where you can't - if a trait has a method which is
*not* `#[track_caller]`, with an impl that *is* `#[track_caller]`, both
the vtable and a function pointer created from that method will be
`ReifyShim(def_id)`.

Currently, the reason is optional to ensure no additional unique
`ReifyShim`s are added without KCFI on. However, the case in which an
extra `ReifyShim` is created is sufficiently rare that this may be worth
revisiting to reduce complexity.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f64da911fcc9844657bb1bc5d0cf08d4dc7d7fbb with merge b04e6b6ae6b129e569e842facca0310b9041704d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f64da911fcc9844657bb1bc5d0cf08d4dc7d7fbb","merge_sha":"b04e6b6ae6b129e569e842facca0310b9041704d"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #123065) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #123071) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
I really don't expect perf to come back with anything other than neutral here, but --

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 4775daf504057aecd3a016b63b9f68c6d9b4f1b2 with merge 0f5b7417e5e3759e1863051a83d15a28103a0e3f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4775daf504057aecd3a016b63b9f68c6d9b4f1b2","merge_sha":"0f5b7417e5e3759e1863051a83d15a28103a0e3f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8503261960/job/23290082852)
Build commit: 0f5b7417e5e3759e1863051a83d15a28103a0e3f (`0f5b7417e5e3759e1863051a83d15a28103a0e3f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8503261960/job/23290082852"},"merge_sha":"0f5b7417e5e3759e1863051a83d15a28103a0e3f"} -->

rust-timer commented:
Queued 0f5b7417e5e3759e1863051a83d15a28103a0e3f with parent 66de6111960571ab7b030c205b6176801c7ecbfc, future [comparison URL](https://perf.rust-lang.org/compare.html?start=66de6111960571ab7b030c205b6176801c7ecbfc&end=0f5b7417e5e3759e1863051a83d15a28103a0e3f).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (0f5b7417e5e3759e1863051a83d15a28103a0e3f): [comparison URL](https://perf.rust-lang.org/compare.html?start=66de6111960571ab7b030c205b6176801c7ecbfc&end=0f5b7417e5e3759e1863051a83d15a28103a0e3f&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=66de6111960571ab7b030c205b6176801c7ecbfc&end=0f5b7417e5e3759e1863051a83d15a28103a0e3f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=66de6111960571ab7b030c205b6176801c7ecbfc&end=0f5b7417e5e3759e1863051a83d15a28103a0e3f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -4.8% | [-4.8%, -4.8%] | 1     |
| All ❌✅ (primary)                 | -1.3% | [-1.3%, -1.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=66de6111960571ab7b030c205b6176801c7ecbfc&end=0f5b7417e5e3759e1863051a83d15a28103a0e3f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [2.1%, 3.6%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.2%, -1.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=66de6111960571ab7b030c205b6176801c7ecbfc&end=0f5b7417e5e3759e1863051a83d15a28103a0e3f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 667.526s -> 668.303s (0.12%)
**Artifact size**: 315.67 MiB -> 315.68 MiB (0.00%)

compiler-errors commented on compiler/rustc_middle/src/ty/instance.rs:
```diff
@@ -31,6 +31,13 @@ pub struct Instance<'tcx> {
     pub args: GenericArgsRef<'tcx>,
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
```
Document this please + the variants specifically. Also note that it's only currently used when KCFI is enabled.

compiler-errors commented on compiler/rustc_symbol_mangling/src/typeid.rs:
```diff
@@ -67,8 +67,13 @@ pub fn kcfi_typeid_for_fnabi<'tcx>(
+    // If we receive a `ReifyShim` intended to produce a function pointer, we need to remain
+    // concrete - abstraction is for vtables.
+    if matches!(instance.def, InstanceDef::ReifyShim(_, Some(ReifyReason::FnPtr))) {
+        options |= TypeIdOptions::NO_SELF_TYPE_ERASURE
```
it feels weird to have a negative option. Totally random ask, but is there a way to flip the polarity here? Maybe have a `default()` on `TypeIdOptions` with `SELF_TYPE_ERASURE` set and then remove it here?

maurer replied:
Shouldn't be too hard, I'll do it in a precursor commit.

compiler-errors commented on compiler/rustc_middle/src/ty/instance.rs:
```diff
@@ -476,15 +491,30 @@ impl<'tcx> Instance<'tcx> {
+                    // KCFI because it can only attach one type per function.
+                    resolved.def = InstanceDef::ReifyShim(resolved.def_id(), reason)
+                }
+                _ if tcx.sess.is_sanitizer_kcfi_enabled()
```
And this one is intercepting what specifically? It's not `|| {} as fn()` casts, right? If it is, I'm specifically curious what the call stack for this is -- `resolve_for_fn_ptr` is only called on `ty::FnDef` afaict.

compiler-errors replied:
Easiest way to get a stack trace is to put `bug!()` here lol

maurer replied:
(This would also end up covering `poll` function pointers, `poll_next` function ponters, and `.next` function pointers for the various flavors of generator/async coroutines)

maurer committed:
CFI: Switch sense of type erasure flag

Previously, we had `NO_SELF_TYPE_ERASURE`, a negative configuration. Now
we have `ERASE_SELF_TYPE`, a positive configuration.

rcvalle commented on compiler/rustc_codegen_llvm/src/declare.rs:
```diff
@@ -147,7 +147,7 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
                     TypeIdOptions::GENERALIZE_POINTERS,
                     TypeIdOptions::NORMALIZE_INTEGERS,
-                    TypeIdOptions::NO_SELF_TYPE_ERASURE,
+                    TypeIdOptions::ERASE_SELF_TYPE,
```
I don't know, I kinda disagree with the change to the NO_SELF_TYPE_ERASURE. I think the default should be what is going to be done most of the time or more commonly, and it's performing type erasure. The exception for the secondary type id should be opted in when needed.

maurer replied:
I don't have a strong preference - I like this version a little better due to the "positive" nature of it.

With regards to defaults vs secondary, only KCFI has a meaningful notion of a "default" type ID - CFI attaches all of them to the same attachment site, so none of them are really primary or secondary. KCFI still sets `ERASE_SELF_TYPE` as its default with this changeset.

That said, if this change is contentious and not required by compiler-errors to land this PR, I don't really care to fight over this.

rcvalle replied:
It's secondary because these are not always included. They're only included when they're unique and for methods, and also are only tested for when methods are used as function pointers so the concept of secondary does apply generally.

rcvalle commented on tests/codegen/sanitizer/cfi/emit-type-metadata-id-itanium-cxx-abi-method-secondary-typeid.rs:
```diff
@@ -18,5 +18,5 @@ impl Trait1 for Type1 {
-// CHECK: ![[TYPE1]] = !{i64 0, !"_ZTSFvu3refIu3dynIu{{[0-9]+}}NtC{{[[:print:]]+}}_{{[[:print:]]+}}6Trait1u6regionEEE"}
-// CHECK: ![[TYPE2]] = !{i64 0, !"_ZTSFvu3refIu{{[0-9]+}}NtC{{[[:print:]]+}}_{{[[:print:]]+}}5Type1EE"}
+// CHECK: ![[TYPE1]] = !{i64 0, !"_ZTSFvu3refIu{{[0-9]+}}NtC{{[[:print:]]+}}_{{[[:print:]]+}}5Type1EE"}
+// CHECK: ![[TYPE2]] = !{i64 0, !"_ZTSFvu3refIu3dynIu{{[0-9]+}}NtC{{[[:print:]]+}}_{{[[:print:]]+}}6Trait1u6regionEEE"}
```
As a result of what I mentioned above, it seems the type id order also changed here.

maurer replied:
Yes. I'm not sure why the type order changing is an issue if we only have one test that checks it.

rcvalle replied:
The tests aren't an issue now. My disagreement is with the less common case (i.e., not performing self type erasure/the secondary type id) being the default behavior and also the first type id.

maurer commented:
@rustbot ready

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 473a70de8457645df7a49558d6ba27405f966ee0 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 473a70de8457645df7a49558d6ba27405f966ee0 -->
<!-- homu: {"type":"Approved","sha":"473a70de8457645df7a49558d6ba27405f966ee0","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 473a70de8457645df7a49558d6ba27405f966ee0 with merge 01940819cf7d17f2cb8161ee2a44083b152ffc1b...
<!-- homu: {"type":"BuildStarted","head_sha":"473a70de8457645df7a49558d6ba27405f966ee0","merge_sha":"01940819cf7d17f2cb8161ee2a44083b152ffc1b"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8546287494/job/23418329390)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8546287494/job/23418329390","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8546287494/job/23418329390) [(plain)](https://github.com/rust-lang-ci/rust/commit/01940819cf7d17f2cb8161ee2a44083b152ffc1b/checks/23418329390/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

compiler-errors commented:
@bors retry

bors commented:
:hourglass: Testing commit 473a70de8457645df7a49558d6ba27405f966ee0 with merge 29fe618f750c5ff7f8fb75871e75280b569b4e67...
<!-- homu: {"type":"BuildStarted","head_sha":"473a70de8457645df7a49558d6ba27405f966ee0","merge_sha":"29fe618f750c5ff7f8fb75871e75280b569b4e67"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8550387921/job/23431478295)
Approved by: compiler-errors
Pushing 29fe618f750c5ff7f8fb75871e75280b569b4e67 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8550387921/job/23431478295"},"merge_sha":"29fe618f750c5ff7f8fb75871e75280b569b4e67"} -->

rust-timer commented:
Finished benchmarking commit (29fe618f750c5ff7f8fb75871e75280b569b4e67): [comparison URL](https://perf.rust-lang.org/compare.html?start=43f4f2a3b1a3d3fb3dbbbe4fde33fb97c780ee98&end=29fe618f750c5ff7f8fb75871e75280b569b4e67&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=43f4f2a3b1a3d3fb3dbbbe4fde33fb97c780ee98&end=29fe618f750c5ff7f8fb75871e75280b569b4e67&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=43f4f2a3b1a3d3fb3dbbbe4fde33fb97c780ee98&end=29fe618f750c5ff7f8fb75871e75280b569b4e67&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.5%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.3%, -1.3%] | 1     |
| All ❌✅ (primary)                 | -0.9% | [-1.5%, -0.1%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=43f4f2a3b1a3d3fb3dbbbe4fde33fb97c780ee98&end=29fe618f750c5ff7f8fb75871e75280b569b4e67&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=43f4f2a3b1a3d3fb3dbbbe4fde33fb97c780ee98&end=29fe618f750c5ff7f8fb75871e75280b569b4e67&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 667.94s -> 669.353s (0.21%)
**Artifact size**: 318.08 MiB -> 318.07 MiB (-0.00%)
